### PR TITLE
ehcache implementation

### PR DESCRIPTION
**Pull request by apurbraj**
## Description

* Implementation of ehcache.
* Add a link to teh design document or include the design bullet points related to this PR here.

_(Ignore if these details are present on the associated JIRA ticket)_

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Design related bullet points or design document link related to this PR added in the description above.

- [ ] Updated corresponding Postman Collection or Api documentation for the changes in this PR.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Add required Swagger annotation and update API documentation with details of any API changes if applicable

- [ ] Followed the naming conventions as given in https://docs.google.com/document/d/1Q4vaMSzrTxxh9TS0RILuNkSkYCxotuYk1Xe0CMIkkCU/edit?usp=sharing

### CODE REVIEW

File: src/main/java/org/mifos/identityaccountmapper/service/AddUpdatePaymentModalityService.java [hunk 1]
```diff
@@ -69,14 +71,53 @@ private void sendCallback(List<ErrorTracking> errorTrackingList, String requestI
...
+    private void validateAndUpdatePaymentModality(List<BeneficiaryDTO> beneficiaryList, RequestDTO request, List<ErrorTracking> errorTrackingList) {
+        beneficiaryList.stream().forEach(beneficiary -> {
+            String requestID = request.getRequestID();
+            Boolean beneficiaryExists = validateBeneficiary(beneficiary, requestID, errorTrackingList);
+            updateModalityDetails(beneficiary, beneficiaryExists, errorTrackingList, requestID, beneficiary.getPayeeIdentity());
+        });
+    }
+
+    @Transactional
+    @CacheEvict(value = "accountLookupCache",key = "#payeeIdentity", allEntries = true)
```
Comment (fynmanoj) [inline src/main/java/org/mifos/identityaccountmapper/service/AddUpdatePaymentModalityService.java:83]: why allEntries is true

File: src/main/java/org/mifos/identityaccountmapper/service/AddUpdatePaymentModalityService.java [hunk 2]
```diff
@@ -69,14 +71,53 @@ private void sendCallback(List<ErrorTracking> errorTrackingList, String requestI
...
             Boolean beneficiaryExists =  validateBeneficiary(beneficiary, requestID, errorTrackingList);
-            try {
-                if(beneficiaryExists){
-                    PaymentModalityDetails paymentModality = fetchPaymentModalityDetails(beneficiary);
+            addModalityDetails(beneficiary, beneficiaryExists, errorTrackingList, requestID, beneficiary.getPayeeIdentity());
+        });
+    }
 
+    @Transactional
+    @CacheEvict(value = "accountLookupCache",key = "#payeeIdentity",  allEntries = true)
```
Comment (fynmanoj) [inline src/main/java/org/mifos/identityaccountmapper/service/AddUpdatePaymentModalityService.java:114]: why allEntries  true

File: src/main/resources/ehcache.xml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                <class>org.mifos.identityaccountmapper.util.CacheEventLogger</class>
+                <event-firing-mode>ASYNCHRONOUS</event-firing-mode>
+                <event-ordering-mode>UNORDERED</event-ordering-mode>
+                <events-to-fire-on>CREATED</events-to-fire-on>
+                <events-to-fire-on>EXPIRED</events-to-fire-on>
+                <events-to-fire-on>EVICTED</events-to-fire-on>
+            </listener>
+        </listeners>
+        <resources>
+            <heap>1000</heap>
```
Comment (fynmanoj) [inline src/main/resources/ehcache.xml:27]: these values should be configurable from deployment

**Review by fynmanoj — CHANGES_REQUESTED** [review]
requires clarification

File: src/main/resources/ehcache.xml (hunk 3 above)
Comment (apurbraj) [reply src/main/resources/ehcache.xml:27]: ehcache.xml file is not supporting placeholders.
Can you suggest any alternate way to make this configurable?

Comment (fynmanoj) [reply src/main/resources/ehcache.xml:27]: https://stackoverflow.com/questions/37301017/accessing-spring-environmental-variables-in-ehcache-xml

**Pull request merged**